In [44]:
x = 5**3

In [45]:
x

125

In [46]:
 y = 2**0.5

In [47]:
y

1.4142135623730951

#### Reassign

In [48]:
x = 7

In [49]:
x

7

In [50]:
x + x

14

In [51]:
x = x + x

In [52]:
x

14

In [53]:
type(x)

int

In [54]:
type(y)

float

In [55]:
s = "hello" 

In [56]:
type(s)

str

In [57]:
l = ["10", "hello"]

In [58]:
type(l)

list

#### A persons income per month is 100. Annual tax rate 12% Calculate total_taxes using variable assignments as much as possible.

In [60]:
# Define variables
monthly_income = 100
months_per_year = 12
annual_income = monthly_income * months_per_year
tax_rate = 0.12  # 12% tax rate

In [61]:
# Calculate total taxes
total_taxes = annual_income * tax_rate

In [62]:
# Print result
print("Total Taxes:", total_taxes)

Total Taxes: 144.0


### Multiple Assignments

In [64]:
a, b, c = 5, 10, 15
x = y = z = 100  # Assigning the same value to multiple variables


### Type Conversion

In [65]:
num = "100"
converted_num = int(num)  # Convert string to integer


In [66]:
converted_num

100

### Working with Constants (Python convention

In [67]:
PI = 3.14159  # Constant (use uppercase names)


In [68]:
PI

3.14159

### Dynamic typing 

In [69]:
x = 5
x = "Now I'm a string"


In [70]:
type(x)

str

### Basic Variable Assignment 

In [71]:
name = "Saji"
age = 25
height = 5.8
is_student = True

print(name, age, height, is_student)


Saji 25 5.8 True



### Changing Variable values 


In [72]:
x = 10
print(x)  # Output: 10

x = 20
print(x)  # Output: 20


10
20


### Assigning Multile Variables at once 

In [73]:
a, b, c = 5, 10, 15
print(a, b, c)  # Output: 5 10 15

x = y = z = 100
print(x, y, z)  # Output: 100 100 100


5 10 15
100 100 100


### Variabe Types

In [74]:
integer_num = 10         # Integer
float_num = 5.5          # Floating point
text = "Hello"           # String
is_python_fun = True     # Boolean (True/False)
complex_num = 2 + 3j     # Complex number

print(type(integer_num))  # Output: <class 'int'>
print(type(float_num))    # Output: <class 'float'>
print(type(text))         # Output: <class 'str'>
print(type(is_python_fun)) # Output: <class 'bool'>
print(type(complex_num))  # Output: <class 'complex'>


<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>
<class 'complex'>


### Concatenation (Combining Strings & Variables)

In [75]:
first_name = "Saji"
last_name = "M"
full_name = first_name + " " + last_name
print(full_name)  # Output: Saji Ali


Saji M


In [76]:
sentence_1 = "Hai" 
gap = " "
sentence_2 = "this is SAJI" 
full_sentence = sentence_1 + gap +sentence_2
print(full_sentence)

Hai this is SAJI


### Mathematicla operation

In [77]:
x = 10
y = 3

sum_result = x + y
difference = x - y
product = x * y
quotient = x / y
remainder = x % y
power = x ** y  # x raised to the power of y

print(sum_result, difference, product, quotient, remainder, power)


13 7 30 3.3333333333333335 1 1000


### Swapping two Variables

In [78]:
a, b = 5, 10
a, b = b, a
print(a, b)  # Output: 10 5


10 5


### Getting data from the user

In [79]:
name = input("Enter your name: ")
age = input("Enter your age: ")

print("Hello", name, "You are", age, "years old.")


Enter your name:  Saji
Enter your age:  28


Hello Saji You are 28 years old.


### Global vs Local Variable 

In [80]:
global_var = "I'm global" # python will automauticall take this as string 

def my_function():
    local_var = "I'm local"
    print(local_var)

my_function()
print(global_var)  # This works
# print(local_var)  # This will give an error because local_var is inside the function


I'm local
I'm global


### Type casting 

In [81]:
x = 5      # Integer
y = "10"   # String

x = str(x)  # Convert to string
y = int(y)  # Convert to integer

print(type(x))  # Output: <class 'str'>
print(type(y))  # Output: <class 'int'>


<class 'str'>
<class 'int'>


### checking Memory Address 

In [82]:
a = 10
b = 10
print(id(a), id(b))  # Might be the same, as Python optimizes small integers


136302994758224 136302994758224


### Mutable vs Immutable 

In [83]:
# Immutable types (cannot be changed)
x = 10   # int
y = "Hello"  # str
z = (1, 2, 3)  # tuple

# Mutable types (can be changed)
my_list = [1, 2, 3]  # list
my_dict = {"name": "Saji", "age": 25}  # dictionary

# Changing mutable types
my_list.append(4)
my_dict["age"] = 26

print(my_list)  # Output: [1, 2, 3, 4]
print(my_dict)  # Output: {'name': 'Saji', 'age': 26}


[1, 2, 3, 4]
{'name': 'Saji', 'age': 26}


### Scope of Variable : Global vs Local 

In [84]:
x = 10  # Global variable

def my_function():
    y = 5  # Local variable
    print(y)  # Accessible inside the function

my_function()
print(x)  # Works
# print(y)  # Error! y is local to the function


5
10


### Changine the global variable with in the fun 

In [87]:
x = 10

def change_global():
    global x  # Modifies global variable
    x = 20
    y = 10
    print(y)

change_global()
print(x)  # Output: 20


10
20


### Dynamic tying (Python flexibility) 

In [89]:
x = 10    # Integer
x = "Hello"  # Now it's a string
x = [1, 2, 3]  # Now it's a list
x = (1, 2, 3)  # Now it's a ?

print(type(x))  # Output: <class 'list'>


<class 'tuple'>


### Variable Aliasing 

In [90]:
a = [1, 2, 3]
b = a  # b points to the same list as a

b.append(4)
print(a)  # Output: [1, 2, 3, 4] (both a and b changed!)


[1, 2, 3, 4]


Memory Explanation

    a = [1, 2, 3]

        A list [1, 2, 3] is created in memory.

        The variable a stores a reference to that list.

    b = a

        b does NOT create a new list.

        Instead, b stores the same reference as a.

        Now, both a and b point to the same list in memory.

    b.append(4)

        Since lists are mutable, modifying b also modifies a because they refer to the same object.

### To over come above (Create an indep copy)

In [92]:
a = [1, 2, 3]
b = a.copy()  # Creates a new list
#or 
b = list(a)

b.append(4)

print(a)  # Output: [1, 2, 3]  (Unchanged)
print(b)  # Output: [1, 2, 3, 4] (Only b changed)


[1, 2, 3]
[1, 2, 3, 4]


# Delete a statement 

In [94]:
x = 100
del x  # Deletes x
# print(x)  # This will cause an error


### Using None

In [95]:
x = None  # Represents an empty value
print(x)  # Output: None


None


### is vs == (compwring Variables) 

In [96]:
a = [1, 2, 3]
b = [1, 2, 3]
c = a #Note : it actuall stores the memory, it doesnt create new list. 

print(a == b)  # True (Values are the same)
print(a is b)  # False (Different memory locations)
print(a is c)  # True (Same memory location)


True
False
True


In [101]:
print(id(a), id(b))  # Different memory addresses


136302902042304 136302902063488


In [97]:
x = 100
y = 100
print(x is y)  # True (same memory location)

x = 1000
y = 1000
print(x is y)  # False (Python creates separate objects)


True
False


✅ Python optimizes small integers (between -5 and 256) by storing them in a shared memory pool.
✅ This means x and y point to the same memory address for values in this range

In [98]:
x = 100
y = 100
print(id(x), id(y))  # Same memory address


136302994949584 136302994949584


In [99]:
x = 1000
y = 1000
print(x is y)  # False


False


Why is this False?

👉 Python does not cache integers above 256, so each assignment creates a new object in memory.
Small integers (-5 to 256) are cached, so variables pointing to the same value will have the same memory address.

Larger numbers (outside the range) are stored as separate objects in memory.

The is operator checks memory location, while == checks value.

### == Vs   is 

In [100]:
x = 1000
y = 1000

print(x == y)  # True (same value)
print(x is y)  # False (different memory locations)


True
False


this optimization only applies to certain immutable objects, specifically:
✅ Cached (Shared Memory)

    Small integers (-5 to 256)

    Interned strings (short strings or identifiers like "hello" but not "a long string")

    Some boolean values (True, False)

    Some small tuples (depending on their contents)

❌ Not Cached (Separate Memory)

    Large integers (> 256 or < -5)

    Floats (1.5, 3.14 → always stored separately)

    Lists ([1, 2, 3] → always stored separately)

    Dictionaries ({"a": 1, "b": 2} → always stored separately)

    Sets & custom objects

🔹 Example for Strings



### Examples for Strings 

In [102]:
a = "hello"
b = "hello"
print(a is b)  # True (small string is cached)

a = "hello world, this is a long string!"
b = "hello world, this is a long string!"
print(a is b)  # False (not cached)


True
False


### Problem 1: List and identity

In [103]:
a = [1, 2, 3]
b = a
b.append(4)

print(a)  # What will be the output?
print(b is a)  # True or False?


[1, 2, 3, 4]
True


### Probem 2: Separate list

In [104]:
x = [10, 20, 30]
y = [10, 20, 30]

print(x == y)  # True or False?
print(x is y)  # True or False?


True
False


### problem 3 String interning 

In [105]:
s1 = "hello"
s2 = "hello"

print(s1 is s2)  # True or False?

s3 = "hello world, this is a long string!"
s4 = "hello world, this is a long string!"

print(s3 is s4)  # True or False?


True
False


### Problem 4: changing mutable object

In [106]:
list1 = [1, 2, 3]
list2 = list1

list2.append(4)

print(list1)  # What will be the output?


[1, 2, 3, 4]


### Proble 5: tuple identity

In [107]:
t1 = (1, 2, 3)
t2 = (1, 2, 3)

print(t1 is t2)  # True or False?


False
